### PR TITLE
Switch to using a channel for the health check

### PR DESCRIPTION
**Pull request by A1Liu**
Switch `HealthService` to use a single polling task that posts to a channel, instead of doing a poll per-listener.

Potential next work:
- use the channel in `waitForActive` and `waitForStatus`

### CODE REVIEW

File: server/utils/channel.ts [hunk 1]
```diff
@@ -57,63 +74,80 @@ export class Channel<T> {
...
+			// some scenario where we destroy to clean-up resources and then crash.
+			// better to just make sure a call to destroy can only happen once
+			return;
 		}
 
+		const error = errorAny
+			? errorAny instanceof Error
+				? errorAny
+				: new Error(errorAny)
+			: undefined;
```
Comment (A1Liu) [inline server/utils/channel.ts:100]: This is probably anti-pattern; I hate that errors in JS are `any` but I guess that's the world we live in

File: server/services/health.ts [hunk 2]
```diff
@@ -1,21 +1,90 @@
...
+					if (!isEqual(previous, health)) {
+						healthsPerService.set(config.location, health);
+						HealthChannel.send(health);
+					}
+				}),
+			);
+
+			for (const result of outcome) {
+				if (result.status === 'rejected') {
+					console.error('Sidekick status check failed:', result.reason);
```
[the remark this answers is not in the document]
Comment (A1Liu) [reply server/services/health.ts]: done

File: server/services/health.ts [hunk 3]
```diff
@@ -1,21 +1,90 @@
...
 import { assertUnreachable, objectKeys } from '../utils/util-types';
 import { ProcessManager } from '../utils/process-manager';
 import { HealthStatus, isActiveStatus } from '../utils/shared-types';
 import { ServiceBuildsService } from './service-builds';
-import EventEmitter from 'events';
-import { Observable } from 'rxjs';
+import { ChannelList } from '../utils/channel';
 
-const emitter = new EventEmitter();
+export const HealthChannel = new ChannelList<ServiceHealthEvent>();
```
[the remark this answers is not in the document]
Comment (A1Liu) [reply server/services/health.ts]: done

File: server/services/health.ts [hunk 4]
```diff
@@ -1,21 +1,90 @@
...
+const healthsPerService = new Map<string, ServiceHealthEvent>();
+healthTask();
+
+async function healthTask() {
+	// keys are service locations
+	let serviceConfigs: ServiceConfig[] = undefined as unknown as any;
+	let serviceConfigWait = Infinity;
+
+	for (let wait = 500; ; ) {
```
[the remark this answers is not in the document]
Comment (A1Liu) [reply server/services/health.ts]: done

File: server/services/health.ts [hunk 5]
```diff
@@ -1,21 +1,90 @@
...
+				}),
+			);
+
+			for (const result of outcome) {
+				if (result.status === 'rejected') {
+					console.error('Sidekick status check failed:', result.reason);
+				}
+			}
+		} catch (e) {
+			console.error(e);
```
[the remark this answers is not in the document]
Comment (A1Liu) [reply server/services/health.ts]: done

**Review by karimsa — APPROVED** [review]
I don't see any major issues, so let's merge and use it for a few days before marking as stable

**Pull request merged**